DDP serialization abstraction

harryadel opened the pull request:
Story time yay!!

During a normal work day @determinds I was called over by some colleagues who had trouble uploading files in a Meteor.js app. They were trying to pass the file using a method but in the server it'd up something like `File {}` completely empty/undefined. Which was very bizarre to them. Mind you these are very seasoned JS developers and not someone who started coding yesterday. My brain fired off immediately and I started thinking of DDP protocol and EJSON and how Meteor passes data to the server and what not but I knew if I were to explain a shred of it to them it'd come off as gibberish so instead I told them to either convert the file to base64 if they still want to pass it using a method or circumvent this whole problem by uploading the file to a [Picker](https://github.com/Meteor-Community-Packages/picker) endpoint akin to uploading in a normal Express application. Needless to say they were both frustrated with Meteor and complained of how quirky it's :sweat_smile: But I tried to convince them it's not.
 
Then I decided to look more into this and how Meteor passes data to the server and then it turns out EJSON can only handle certain types, sure we could create a workaround by converting files internally to base64 so it can be string-ified and passed but it felt like a band aid so I started researching more into libraries similar to EJSON that can fulfill this requirement. I stumbled upon MessagePack, then I remembered a recent [thread](https://forums.meteor.com/t/what-if-meteor-was-created-in-2025/63566?u=harry97) that talked about something called bor or cobor?? and well the rest is history.

CBOR is not only more performant but also supports many more types so we kinda get a two for one deal here. Instead of building from scratch, [cbor-x](https://github.com/kriszyp/cbor-x) is used.  This is a backwards-compatible change, all of the APIs exported previously by EJSON are maintained. `METEOR_DDP_CBOR_BINARY` can be enabled where it'd use cbor + base64 as sockjs only allows text based communication for maximum backwards compatibility. 

I don't know of a way to link packages internally so I opted to release a new package [`harry97:cbor`](https://packosphere.com/harry97/cbor). mongo-id adds a new type so it had to be forked and published under `harry97:mongo-id`. Once you're ok with these changes you can of course modify the naming within core packages and publish the official Meteor versions of them. You can pull this PR and run core packages like so to verify it doesn't break anything: `meteor test-packages ./packages/logging`

So in a sense this PR started off to figure out a way to pass files in Meteor methods but accidentally replaces EJSON with CBOR? I'm definitively interested in hearing your opinions. Let's get this going :+1: 

----------

EDIT: This PR now is much more simpler focusing on abstracting the serialization layer without overhauling the entire Meteor framework in the process :sweat_smile:

harryadel committed:
Add cbor package

harryadel committed:
[diff-sequence] Replace EJSON with CBOR

harryadel committed:
[oauth-encryption] Replace EJSON with CBOR

harryadel committed:
[accounts-password] Remove ejson

harryadel committed:
[accounts-passwordless] Remove ejson package

harryadel committed:
[allow-deny] Replace EJSON with CBOR

harryadel committed:
[accounts-base] Replace EJSON with CBOR

harryadel committed:
[browser-policy] Replace EJSON with CBOR

harryadel committed:
[mongo-id] Replace EJSON with CBOR

harryadel committed:
[ddp-client] Replace EJSON with CBOR

harryadel committed:
[base64] Replace EJSON with CBOR

harryadel committed:
[ddp-server] Replace EJSON with CBOR

harryadel committed:
[ddp-common] Replace EJSON with CBOR

harryadel committed:
[check] Replace EJSON with CBOR

harryadel committed:
[minimongo] Replace EJSON with CBOR

harryadel committed:
[ejson] Move to deprecated directory

harryadel committed:
[id-map] Replace EJSON with CBOR

harryadel committed:
[mongo-decimal] Replace EJSON with CBOR

harryadel committed:
[cbor] Remove check package

harryadel committed:
[mongo] Replace EJSON with CBOR

harryadel committed:
[logging] Replace EJSON with CBOR

harryadel committed:
[test-helpers] Replace EJSON with CBOR

harryadel committed:
[test-in-console] Replace EJSON with CBOR

harryadel committed:
[reactive-dict] Replace EJSON with CBOR

harryadel committed:
[test-server-tests-in-console-once] Replace EJSON with CBOR

harryadel committed:
[tinytest] Replace EJSON with CBOR

harryadel committed:
Merge branch 'devel' of github.com:harryadel/meteor into cbor

harryadel committed:
[ddp-common] Remove unused methods

harryadel committed:
[cbor] Fix missing methods

harryadel committed:
Document METEOR_DDP_CBOR_BINARY

italojs commented:
loved, waiting for the tests to pass

Could you bring a few references about CBOR and a few lists of new capabilities we will gain with it (push files, for example)?

harryadel commented:
Tests won't really pass unless you publish cbor under meteor.

EDIT: there're couple of problems with getting tests to work, I'll try to move things as much as possible then report back.

On Monday, October 13, 2025, Italo José ***@***.***> wrote:

> *italojs* left a comment (meteor/meteor#13980)
> <https://github.com/meteor/meteor/pull/13980#issuecomment-3397787600>
>
> loved, waiting for the tests to pass
>
> Could you bring a few references about CBOR and a few lists of new
> capabilities we will gain with it (push files, for example)?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/meteor/meteor/pull/13980#issuecomment-3397787600>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSWGQNDM6NBPJLB3GTYBQL3XOZ7JAVCNFSM6AAAAACI6ZE5GOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGOJXG44DONRQGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

harryadel committed:
[cbor] Move it outside of core packages

This was blocking tests from running

harryadel committed:
Merge branch 'devel' into cbor

harryadel committed:
Utilize native websockets if possible

harryadel committed:
update cbor version

harryadel committed:
Update mongo-id version

harryadel committed:
[tinytest] Replace ejson with cbor

harryadel committed:
[session] Remove ejson

harryadel committed:
[mongo-decimal] Replace ejson with cbor

harryadel committed:
[cbor] Move cbor package back into the core

harryadel committed:
Comment out ejson from isopackets for now

nachocodoner commented:
Cool to see this, @harryadel. Great experimantation! I agree having an advanced way to treat all types, also files, is great to have.

EJSON is used widely across the Meteor runtime, and in the past I looked for ways to make it faster, so relying on a more optimized approach is great. We need tests passing, since this core change touches many places, and we should measure its impact on Meteor 3 performance. It also looks like a good fit for Meteor 3.5, where a lot of benchmarking is happening. Once benchmarks are fixed for the next Change Stream server changes, we can measure how the new encoding strategies affect them, get insight into the gains, and use that to verify and consolidate the change.

harryadel commented:
@nachocodoner Hello Nacho, thanks for chiming in and being open to experimentation. I'm just trying different things and following curiosity/creativity. I know this touches many parts of the Meteor core so we're ought to ensure this doesn't break anything. so no worries I'm in it for the long run although I know how tiresome it'd be :sweat_smile:  
For now getting tests to run pass with this weird packages structure is the goal. ejson moved to deprecated, cbor added, mongo-id and mongo-decimal too were modified :man_shrugging:

harryadel committed:
Replace ejson with harry97:cbor in tools directory

harryadel committed:
Re-enlist original mongo-id changes

harryadel committed:
[cbor] Fix customtype check

harryadel committed:
[mongo] fix test

radekmie commented:
I'd like to raise a small concern regarding this PR: is Meteor Software _really_ planning to replace the underlying data format without a transition period? This is a **huge** deal, since rolling out this change will mean that **all** existing connections will simply break, as after they connect to the new server, they won't be able to receive the information about the new version being available (it'll be sent to them in a different format).

I see there is _some_ fallback in `DDPCommon.parseDDP`, but - correct me if I'm wrong - it will **not** work for the custom types registered via `EJSON.addType` (including `Date`).

I'm not against experimenting with new data formats, but I think there should be a distinction between the data format used internally (e.g., when cloning method parameters) and the one used on the wire (which can be replaced with patching `DDPCommon.parseDDP` and `DDPCommon.stringifyDDP`; plus a small core-level tweak to allow for binary formats if needed).

Another concern is that the `cbor-x` package has literally 1 (one!) maintainer. And I don't want to be a fearmonger here, but after the recent supply chain attacks, I'm more and more sceptical of relying on such (and I know for a fact that "big companies" are too). There are also [quite a few unaddressed issues](https://github.com/kriszyp/cbor-x/issues), including https://github.com/kriszyp/cbor-x/issues/117, which will be an issue for applications with a strict CSP.

harryadel commented:
It's always a pleasure to hear from you @radekmie and I'm personally glad to get to learn from you. The Meteor community is lucky to have you around.

> Is Meteor Software really planning to replace the underlying data format without a transition period?

I think @nachocodoner @italojs @Grubba27 are the authority here but if we you were to read the previous comments made by @nachocodoner and me. This is nothing but an experiment really that both of us are happy to see through. We've not committed to anything yet really. Heck there're few failing tests that I'd still have to fix before asking for a proper review really! So yeah this is more of a public experiment to gather feedback and brain storm together about Meteor problems and solutions. I explained my motives for this PR more than the nitty gritty of code. It's important to consider the current status of Meteor and how newcomers interact with it. Meteor needs to create a ["pit of success"](https://blog.codinghorror.com/falling-into-the-pit-of-success/).

This PR originated from a specific problem . i.e. How can we pass Files to methods? This was my way of answering this question, if you can offer a better way or suggest a way to improve the current way I'll gladly oblige.

> This is a huge deal, since rolling out this change will mean that all existing connections will simply break, as after they connect to the new server, they won't be able to receive the information about the new version being available (it'll be sent to them in a different format).

This is definitely a interesting point I've not thought of, how'd existing connections utilize the new format. Would it cause any breakage or interrupting of flow? This must be addressed for sure. Though this is what partially motivated me to create a backwards compatible package with 1:1 similar API to ensure no breakage. 

>I see there is some fallback in DDPCommon.parseDDP, but - correct me if I'm wrong - it will not work for the custom types registered via EJSON.addType (including Date).

Actually custom types are supported in a backwards compatible way.

 >Another concern is that the cbor-x package has literally 1 (one!) maintainer. And I don't want to be a fearmonger here, but after the recent supply chain attacks, I'm more and more sceptical of relying on such (and I know for a fact that "big companies" are too). There are also quite a few unaddressed issues, including kriszyp/cbor-x#117, which will be an issue for applications with a strict CSP.
 
 Thank you for pointing this out. I didn't consider this aspect when picking a library. I only hoped for a quick library that'll grant Meteor performance improvements. Maybe a copy-fork will help against it? I dunno.
 
 
 All in all, thank for weighing in. Remaining tests are ought to be fixed and few preparations made then hopefully this can be revisited again.

radekmie commented:
> [...] but if we you were to read the previous comments [...] This is nothing but an experiment really that both of us are happy to see through. We've not committed to anything yet really.

I read them. But then I see the labels and a milestone added, which suggests the next release (I assume 3.4 is already closed feature-wise). I know there are no "experiment-specific" labels, but the tone of the comments suggests "we have to measure the performance impact" as the only obstacle. (I guess that's not the case, but that's the tone.)

> How can we pass Files to methods? This was my way of answering this question, if you can offer a better way or suggest a way to improve the current way I'll gladly oblige.

To be perfectly honest, sending _files_ (or any binary blobs) via a WebSocket is not a good idea performance-wise. First of all, there's an inevitable memory limit on the file size. Secondly, there's no way to parallelize the upload, do a partial one, or resume it after going offline. And possible operations on the file blob on the server are also limited (usually, you will either push it to some storage or parse straight away).

Allowing the client to upload files directly to some storage (e.g., an S3-compatible one) is strictly superior in all cases except for the ease of use. (It's not limited to WebSockets; the same applies to all kinds of APIs, including REST or GraphQL.)

But assuming we _really_ want to make it happen, I'd either accept the 33% cost (base64 transformation), or focus on supporting binary blobs in the DDP (EJSON already supports it, so it's just about transferring these more efficiently).

A non-trivial, but also non-invasive way of sending them would be sending them in dedicated follow-up binary messages. Just like the [Transferable objects](https://html.spec.whatwg.org/multipage/structured-data.html#transferable-objects), we could somehow replace the binary blob with some placeholder, and send them right after their original message.

> Though this is what partially motivated me to create a backwards compatible package with 1:1 similar API to ensure no breakage.
>
> Actually custom types are supported in a backwards compatible way.

API-level incompatibility is one topic (it's not fully addressed yet, as the package name has changed), but the wire format is another. At least for the release, the server should be able to accept both formats _and_ negotiate which one to send the messages in.

This could be a good case for a new DDP version. Whether it'll be a new wire format, or a support for binary blobs, DDP already has a version negotiation logic which could be used here with preserving backward compatibility.

> Maybe a copy-fork will help against it? I dunno.

Maybe. But then it comes at a much higher cost of maintaining a whole new library.

nachocodoner commented:
This is definitely considered a contribution. Contributions are outside the official roadmap.  We evaluate them when it's time to plan releases. If it's a big change like this one, we appreciate early feedback and extra focus when we move into the next phase. We tagged it for 3.5 so we don't lose track of this PR, and so we revisit it often, at least for the next release. But from experience, contributions often slide to later releases. We usually don't feel comfortable merging some of them right away if they involve possible breaking changes or performance risks, and we want more evidence, and community and core time put into them. This PR is in so early phase. Maybe an experimentation flag would be more suitable, but so far our process has been: keep contributions visible, revisit them, and try to eventually ship them, that is the reason to append them in release tracks. But to avoid confusion and release expectations we may add a new experimental flag. I'm going to discuss with the team.

Regarding this PR, we need to measure performance impact surely, gather feedback and deep analysis from our team, and also from the community like you're doing, which we appreciate a lot. We also need time to look for other concerns in this area. I understand this changes the foundation of how we handle data. It may keep API compatibility, but it could still break things in other ways as you stated. Because of that, I think this PR needs more work: full testing cycles, betas, possible release candidates if it gets that far, and also a fast way to revert it.

Whatever the final answer is, either replacing the current approach or extending EJSON to handle effeciently binary blobs in files, what I like here is the spark of experimentation and action. I prefer to have a PR that actually changes things instead of long discussions in forums or GitHub without action. I want this to push everyone to discuss what's best. We appreciate the feedback and experience here.

@harryadel if you want to keep exploring this approach, that's completely fine, or you can look into the alternatives mentioned. In either case, I would also suggest a way to make this opt-in behind an experimental config (like `modern` for bundler). I know can be tricky to achieve for a EJSON replacement. With that opt-in setting, we can already achieve easily to introduce no breaking change, and a fallback mechanism to EJSON. Also, if we find a good path, this could become a community package to expand usage and discussion, and maybe reach core later with more consensus.

Let's keep the creative and innovative process going.

nachocodoner commented:
> But assuming we really want to make it happen, I'd either accept the 33% cost (base64 transformation), or focus on supporting binary blobs in the DDP (EJSON already supports it, so it's just about transferring these more efficiently).

On the topic itself, I believe this is the most straightforward approach to reduce the breaking changes we discussed. This is how I upload images in my apps: I convert them to base64 and send them through the method. I understand there are performance and capacity concerns, but in the cases I've tried it hasn't been that bad. Maybe this is the minimum we can ship. Having an opt-in mechanism in existing methods that can detect files and do the transformation in the method API, without pushing that work to Meteor app devs, would be great. Or we can try to orchestrate sending binary blobs for bigger scale.

This doesn't block us from brainstorming and experimenting with the other approach. It can happen that after a few iterations on what is hard, it becomes a strong candidate in the future, maybe also giving a performance boost or other wins. Just that starting with something small and achievable in the short term helps you gain experience with it and already gives people a usable path.

Anyone is welcome to contribute and open a PR to address these points. Regarding the core, work on DDP is already in progress, and we’ll also take into account everything discussed here to explore new complementary directions. Having multiple approaches to compare will give us clearer insights and more direct answers. In my opinion, the key is to always keep options open for everyone who wants to contribute instead of sticking to a single approach defined by the Core team.

harryadel commented:
Not to toot my own horn but thank you @nachocodoner <3. Talk is cheap and lots of people talk. I'd rather start a PR, discuss it and close it rather than just talk and talk. Surely, it's  a bit work but I get to learn and there's work in public that others can rely on later in the future to avoid pitfalls or even improve the proposed solution. Shoot first, ask later :sweat_smile: 

There's a value in EJSON to CBOR conversion regardless of this PR.

PROS:
- Native support for alternative types (including Files but not exclusively) 
- Potential Performance gains? whether in encoding or transmission 
- Modern approach

CONS:
- Potential breaking changes to current apps 
- Supply chain attacks given our reliance on external implementation of CBOR instead of our own

Since what's left to get the tests passing is very little compared to rolling everything back I'm going to continue and then see what the future holds for us. 

 @radekmie Thank you for chiming in, contesting ideas is great for the overall health of our community and our beloved Meteor. :pray:

italojs commented:
@radekmie Always brings good points. I like it when he reviews the PRs. We won't consider those points if he didn't bring them. I understand Radek's point when he mentioned the labels and milestones. Maybe we should have better indicators of the real state of the PR. We never merge critical changes like this without a thorough analysis of security, DX, performance, and other factors.
At the same time, the point about experiment -> discuss experiment again is the exact behavior we love to see.

What we need here is a guide for the next steps. The PR shows a good promise; we should go ahead, but I see a few questions that need answers first:

- [x] Ensure it solves the issue 
- [ ] Ensure it has a fallback 
- [ ] Discuss the [cbor-x](https://github.com/kriszyp/cbor-x) usage 
    * [ ]     Do we have another alternative?
    * [ ]     Can we maintain it?
    * [ ]     How much effort do we need to solve the main issues(and who will do it)?
- [ ] - Measure how much faster it is compared with EJSON and fallback-EJSON

Feel free to contribute to this list, guys ;)

@harryadel I have an unfinished project where I use [OpenTelemetry](https://opentelemetry.io/) + [Prometheus](https://prometheus.io/) + [Grafana](https://grafana.com/) to measure a few metrics from meteorjs app, I will try to push it beween today or tomorrow, then we weill be able to mensure the performance of it, in this mean time you can work on the other questions

italojs commented:
@harryadel  [Here](https://github.com/meteor/performance/tree/otel/otel) inside `/otel` folder you can setup a few benchmarks

```
# shell1
meteor npm i
docker compose up -d
npm run start
```

```
# shell2
npx artillery run tests/artillery/add-task.yml   # it just test the server
```
Open Grafana at localhost:3000 

<img width="1498" height="747" alt="Screenshot 2025-10-27 at 4 01 33 PM" src="https://github.com/user-attachments/assets/53437feb-8f08-45e2-8e97-6a41a964c9b0" />

You can see a few metrics inside `dashabords` menu session (PS: Dont trust on `trace` numbers from dashobards`

mvogttech commented:
Original author of the "Building Meteor in 2025" here.

It's exciting to see Meteor's modernization shifting from forum discussions to active PRs great work on pushing this forward @harryadel!

From my experiments earlier this year, the optimal path is to prioritize refactoring the core DDP and EJSON logic for extensibility via abstractions before diving into specific integrations like CBOR. This keeps DDP as a protocol-agnostic communication standard, decoupling it from specific transports (e.g., the current SockJS/WebSocket setup) and serialization formats (e.g., EJSON).

Here are refined steps for setting up Meteor to integrate newer serialization libraries, drawing from the current DDP implementation.

## 1. Decouple DDP from Transport and Serialization
Currently, DDP on the client uses a `LivedataConnection` class (exposed as `DDP.Connection`) that tightly couples to a `Stream` transport (e.g., SockJS for WebSockets) and EJSON for message handling. On the server, similar logic in `livedata_server.js` processes messages via EJSON.parse and handles connections through a conduit. We can abstract this by defining a core `DDP` class with required methods for transports and serializers to implement, ensuring the core focuses on protocol semantics like methods, subs, and data updates.

Example abstraction, inspired by existing `LivedataConnection` structure:
```
// packages/ddp-core/abstract/ddp.js
import { EJSON } from 'meteor/ejson'; // For fallback or registry

export class DDP {
  constructor(options = {}) {
    this.transport = options.transport; // Injected transport instance (e.g., SockJSStream)
    this.serializer = options.serializer || EJSON; // Default to EJSON
    this.registry = {
      transports: new Map(),
      serializers: new Map()
    };
    this._methodInvokers = {}; // Track outstanding methods
    this._subs = {}; // Track subscriptions
    // ... other state from current LivedataConnection
  }

  // Abstract connect, delegated to transport
  async connect(url, retry = true) {
    if (!this.transport) throw new Error('Transport not set');
    await this.transport.connect(url);
    this.transport.on('message', (raw) => this._processMessage(raw));
    this.transport.on('reset', () => this._handleReset());
    // Handle reconnection logic if retry=true, mirroring current auto-reconnect
  }

  // Send message, using serializer and transport
  send(message) {
    const serialized = this.serializer.stringify(message);
    this.transport.send(serialized);
  }

  // Process incoming, using serializer
  _processMessage(raw) {
    let msg;
    try {
      msg = this.serializer.parse(raw);
    } catch (e) {
      console.error('Parse error:', e);
      return;
    }
    // Dispatch based on msg.msg: 'method', 'result', 'updated', 'added', etc.
    if (msg.msg === 'result') {
      this._handleMethodResult(msg);
    } else if (['added', 'changed', 'removed'].includes(msg.msg)) {
      this._handleDataUpdate(msg);
    } // ... extend for other types like 'ready', 'nosub'
  }

  // Registry methods for plug-and-play
  registerTransport(name, transportClass) {
    this.registry.transports.set(name, transportClass);
  }

  registerSerializer(name, serializer) {
    this.registry.serializers.set(name, serializer); // { stringify, parse }
  }

  // Negotiation during connect handshake
  negotiate(clientCapabilities) {
    // Example: Client sends {transports: ['websocket', 'webtransport'], serializers: ['ejson', 'cbor']}
    const preferredTransport = this._selectBest(clientCapabilities.transports, this.registry.transports);
    const preferredSerializer = this._selectBest(clientCapabilities.serializers, this.registry.serializers);
    this.transport = new (this.registry.transports.get(preferredTransport))();
    this.serializer = this.registry.serializers.get(preferredSerializer);
    return { selectedTransport: preferredTransport, selectedSerializer: preferredSerializer };
  }

  // Core handlers (agnostic)
  _handleMethodResult(msg) {
    const invoker = this._methodInvokers[msg.id];
    if (invoker) {
      if (msg.error) invoker.setError(msg.error);
      else invoker.setResult(msg.result);
      delete this._methodInvokers[msg.id];
    }
  }

  _handleDataUpdate(msg) {
    // Update local stores or trigger reactivity, as in current _livedata_data
  }
}
```
This mirrors the existing client-side flow (e.g., `_stream.send` and `EJSON.parse` in `livedata_connection.js`) but makes transport and serializer injectable, reducing coupling.

## 2. Implement a Registry for Transports and Serializers
Build a central registry in `ddp-core` for dynamic registration. Current implementations (e.g., SockJS stream on client, WebSocket on server) can auto-register as defaults.

Example:
```
// In ddp-websocket package
import { DDP } from 'meteor/ddp-core';
import SockJS from 'sockjs-client'; // Or native WebSocket

class WebSocketTransport {
  constructor(url) {
    this._sock = new SockJS(url); // Or WebSocket(url)
  }
  async connect() {
    return new Promise((resolve, reject) => {
      this._sock.onopen = resolve;
      this._sock.onerror = reject;
    });
  }
  send(data) {
    this._sock.send(data);
  }
  on(event, callback) {
    if (event === 'message') this._sock.onmessage = callback;
    // ... other events like 'close', 'error'
  }
}

DDP.registerTransport('websocket', WebSocketTransport);
```

For serializers:
```
// In ddp-cbor package
const CBOR = require('cbor-x');

const CborSerializer = {
  stringify: (obj) => CBOR.encode(obj),
  parse: (data) => CBOR.decode(data)
};

DDP.registerSerializer('cbor', CborSerializer);
```

## 3. Migrate Existing Implementations
Extract current logic to separate packages: `ddp-websocket` for transport (based on `stream_client_sockjs.js`), `ddp-ejson` for serialization (wrapping `ejson.js`'s stringify/parse). Auto-register on import for compatibility—no changes for existing apps.

## 4. Add Client-Server Negotiation
Extend the 'connect' DDP message with capabilities, as in current handshake (version negotiation via 'support' field). Server responds with selected options.

Example extended message:
```
{ msg: 'connect', version: '1', support: ['1', 'pre2'], capabilities: { transports: ['websocket', 'webtransport'], serializers: ['ejson', 'cbor'] } }
```

## 5. Test and Integrate New Options
With the framework, adding CBOR is simple: Implement the serializer, register it, and benchmark.

This sets Meteor up for future innovations in protocol communications and data serialization formats without core rewrites.

Relevant to this discussion: [JSON Serialization Formats](https://jsonjoy.com/blog/json-codec-benchmarks).

radekmie commented:
@harryadel

> Native support for alternative types (including Files but not exclusively)

Just for the record: it can be achieved now using `EJSON.addType`. You could even create a small Meteor package which would register it on both client and server with respective base64 conversion for the transfer.

> Potential breaking changes to current apps

I'll keep being pedantic about it: this **is**, not **may be**, a breaking change.

---

@italojs

> Maybe we should have better indicators of the real state of the PR.

The `in-discussion` seem good enough.

---

@mvogttech

> Extend the 'connect' DDP message with capabilities, as in current handshake (version negotiation via 'support' field). Server responds with selected options.

While the [DDP specification](https://github.com/meteor/meteor/blob/devel/packages/ddp/DDP.md) is "[...] a rough description of the protocol and not intended to be entirely definitive [...]", I'd like to highlight that there are a few DDP non-JS implementations out there, and for some of them (e.g., the one in the [DDP Router](https://github.com/meteor/ddp-router), sending additional fields _will_ be a breaking change on its own.)

An alternative would be to do something like "scoped versions", e.g., `2.cbor`. Please note that the first message (`connect`) should stick to EJSON, at least for the time being. I also believe the new formats should be opt-in, so it's possible for every connected client to opt-in.

mvogttech commented:
> An alternative would be to do something like "scoped versions", e.g., `2.cbor`. Please note that the first message (`connect`) should stick to EJSON, at least for the time being. I also believe the new formats should be opt-in, so it's possible for every connected client to opt-in.

I agree.

This is the exact design I was trying to convey with my previous message.

harryadel committed:
Merge branch 'devel' into cbor

harryadel committed:
fix DDP/CBOR type compatibility and null-safety

- Guard Object.keys() call against null msg in ConnectionStreamHandlers
- Remove '1.cbor' from SUPPORTED_DDP_VERSIONS until full negotiation
  infrastructure is in place
- Restore CBOR._adjustTypes*/clone in wireToAbstract/abstractToWire
- Fix CBOR type adjusters to be EJSON wire-format compatible:
  - Add $InfNaN encoding/decoding for Infinity, -Infinity, NaN
  - Add $literal wrapping for plain objects with $-prefixed keys
  - Fix $date: 0 (epoch) not recognized due to falsy check

harryadel committed:
wire up negotiated serializer in DDP client and server send/receive

harryadel committed:
Revert CBOR-for-EJSON replacement, keep DDP serialization abstraction

Restore EJSON as the canonical type system across all packages.
Remove the harry97:cbor Meteor package and revert all CBOR.* calls
back to EJSON.* in non-DDP packages.

The DDP serialization abstraction layer (SerializerRegistry, pluggable
JSON/CBOR serializers, version negotiation) is preserved in ddp-common,
now backed directly by the cbor-x npm package instead of the removed
Meteor package.

harryadel committed:
Fix broken file

harryadel committed:
Revert changes made to packages/minimongo/minimongo_tests_client.js

harryadel committed:
Remove non-DDP changes to keep PR focused on serialization abstraction

Revert EJSON import additions and package.js cleanups that are
unrelated to the DDP serialization layer.

harryadel committed:
Merge branch 'devel' into cbor

harryadel committed:
Remove CBOR serializer from this PR, keep only the abstraction layer

Strip cbor-x dependency, CBOR serializer registration, and 1.cbor
version mapping. The SerializerRegistry and version negotiation
infrastructure remain — adding a binary serializer (CBOR, BSON, etc.)
is a follow-up decision for the community.

harryadel committed:
Remove binary transport changes, keep only serialization abstraction

Revert socket-stream-client binary support (_supportsBinary,
binaryType, native WebSocket preference) and remove supportsBinary
option passing — these belong in the PR that adds a binary serializer.

harryadel committed:
Remove unrelated blaze submodule change

harryadel commented:
Thank you guys for your comments. It made me realize the fault in my ways. I was trying to rip out EJSON from all Meteor packages in one go, along with irreversible change to the DDP protocol as icing on the cake :grin: stubbornness of the youth lol 

Also the recent developments by the community in https://github.com/meteor/meteor/pull/14190 and https://github.com/meteor/meteor/pull/14206 made me rethink the changes further.

So I opted for a lot more toned down version that lays the ground work for future work (either by me or the community) where we abstract the serialization layer. This way DDP is on its way to becoming truly an agnostic protocol. Hopefully even the DB layer :crossed_fingers: 

In later PRs we can decide on the best serialization layer and how to best integrate it. 

Awaiting your feedback guys, thanks!
@radekmie @mvogttech @italojs @nachocodoner

radekmie reviewed (commented):
A few comments from my end:
* I **really** like that it's now opt-in.
* I believe `parseDDP` should take an options object, just like `stringifyDDP`. It's common to share the same options between the parsers and serializers, e.g., to configure dictionaries or compression level.
    * I also think `format` would be a better name, as it also implies the parser, not only the serializer.
* It'd be safer to make the `DDPCommon.SerializerRegistry.get` fallback to `json` automatically, without the need of ensuring the format in `parseDDP`/`stringifyDDP`. That's the only option to make it truly backwards-compatible.
* And of course, we need tests 😅

italojs commented on packages/ddp-client/common/livedata_connection.js:
```diff
@@ -95,6 +95,7 @@ export class Connection {
     self._lastSessionId = null;
     self._versionSuggestion = null; // The last proposed DDP version.
     self._version = null; // The DDP version agreed on by client and server.
+    self._serializer = 'json'; // Negotiated serializer, switches after 'connected'.
```
Probability tests can answer my question, but how the dev will update this value in runtime since it's "private"

italojs commented on packages/ddp-common/utils.js:
```diff
@@ -1,5 +1,7 @@
 "use strict";
 
+import { EJSON } from 'meteor/ejson';
```
EJSON is global, we can see a global usage [here](https://github.com/meteor/meteor/blob/86a32d99841e682fc23e52bb58f30ee0e8a88fc1/packages/accounts-base/accounts_server.js#L1616)

harryadel commented:
Closing in favor of https://github.com/meteor/meteor/pull/14235
